[Do not merge] Experimenter update feature

btruhand opened the pull request:
### To do
- Front end for experimenters to easily update their data in the database

### Done
- Backend logic added for updating experimenter's data

jorgearanda commented on app.js:
```diff
@@ -113,6 +113,11 @@ app.get('/runs', isUser, runs.list);
+// app.put('/experimenters/:id', isUser, experimenters.update);
+
+// Route to use for testing
+app.put('/experimenters/:id', experimenters.update);
```
OK--but let's switch to the route with middleware as soon as possible. We should make a check on :id being equal to the session user id--see below.

jorgearanda commented on routes/experimenters.js:
```diff
@@ -30,3 +31,39 @@ exports.create = function (req, res) {
    });
 };
+
+function updateExperimenter(req, res, exp) {
```
Let's put this function after the exports.update declaration, to keep the exports together.

jorgearanda commented on routes/experimenters.js:
```diff
@@ -30,3 +31,39 @@ exports.create = function (req, res) {
+           return res.send(400);
+       }
+
+       return res.status(200).send(rawRes);
```
For successful PUT requests I prefer to send a 204 without content, like so:
    return res.send(204);

jorgearanda commented on routes/experimenters.js:
```diff
@@ -30,3 +31,39 @@ exports.create = function (req, res) {
+
+// PUT /experimenters/:id
+exports.update = function(req, res) {
+    var exp = {};
```
Early on, check that the request is for the user's own resource: `if (req.params.id !== req.session.userId) return res.send(401);`

(This is assuming the user id is at req.session.userId—please check whether that's the case)

jorgearanda commented on routes/experimenters.js:
```diff
@@ -30,3 +31,39 @@ exports.create = function (req, res) {
+    var exp = {};
+    if(req.body.username !== undefined) exp.username = req.body.username;
+    if(req.body.name !== undefined) exp.name = req.body.name;
+    if(req.body.email !== undefined) exp.email = req.body.email;
```
I prefer checking for truthy values on these: `if (req.body.username) exp.username = req.body.username;`

jorgearanda commented on routes/runs.js:
```diff
@@ -46,7 +46,7 @@ function flattenRunResults(runs) {
 
 // Generates a CSV file containing the results of the queried runs
-function generateCSVRuns(runs, res) {
+function generateCSVRuns(runs, res, req) {
```
By convention, let's send `req` before `res` always

btruhand commented:
Hi Jorge,

I was wondering how you'd like the front end for the experimenters to update their data to be like

jorgearanda commented:
Let's add a button, "Change my User Profile", at the end of the experimenter's dashboard page. The button takes the experimenter into a different page, in which they get a form to change their profile information.

btruhand commented:
After they submit the form, where will the users be taken to?

jorgearanda commented:
Back to the dashboard

btruhand commented:
So I've added the front end so that experimenters can update their data. Tell me what you think of it

Also I was wondering what should happen in the case of the update not going smoothly i.e it couldn't update their data probably. Should it stay in the same page indicating that there was some kind of error? Or should it go back to the dashboard indicating that there was some kind of error (and if it does the latter should the users be notified if the update is a success or not)?

jorgearanda commented:
In the case of error, we should notify the user on the same editing screen. We should always offer an option to Cancel and go back to the dashboard. And thanks for the changes--I'll take a look at them later!

jorgearanda commented:
Nice, Btara, thanks! A couple of minor things: could you please put the button at the end of the dashboard, rather than at the beginning? It should be a rarely used feature. And could you add a Cancel button on the new page?

btruhand commented:
I've moved the button at the end of the dashboard and I've added a cancel button to the new page. I've done some error handling in the server side (in case people turn off javascript on the client side), though I've only handled the case when none of the fields are filled and verifying email format so that it looks like a legit email (though not checking the domain).

I only added an alert box for handling the errors and a general error message. Tell me if you want me to handle more specific cases and output more specific error message for each kind

jorgearanda commented:
Thanks Btara--I'll take a look at this tomorrow.

jorgearanda commented on public/js/update-experimenter.js:
```diff
@@ -0,0 +1,37 @@
+'use strict';
```
One thing to add is the current user profile data, so that we don't land into a page with a blank form

jorgearanda replied:
Of course, for the password that's impossible to do :-)

jorgearanda commented on public/js/update-experimenter.js:
```diff
@@ -0,0 +1,37 @@
+
+function submitUpdate() {
+    var name = $('#name').val();
+    var username = $('#username').val();
```
I've been thinking about letting users change their username, and I think at least for now it's best not to allow this. It requires us to propagate the change onto microworlds and runs, and it should be a very uncommon thing to do anyway. So let's please take out the option to do so through this form.

jorgearanda commented on views/update-experimenter.jade:
```diff
@@ -0,0 +1,37 @@
+            label.form-label.col-md-6(for='password') Password
+            .col-md-6
+              input#password.form-control(type='text', name='rawPassword')
+        button#do-update.btn.btn-info(type='submit') Update profile
```
It's best to adhere to the convention in the rest of the system for the OK/Cancel buttons: OK (in this case, Update profile) on green, and Cancel without a background--see for instance the page for creating a new microworld.

jorgearanda commented on views/update-experimenter.jade:
```diff
@@ -0,0 +1,37 @@
+              input#email.form-control(type='email', name='email')
+        .row
+          .form-group
+            label.form-label.col-md-6(for='password') Password
```
Make this of type password so the text is not visible

jorgearanda replied:
(and let's have the user "Confirm" the password a second time as well)

jorgearanda commented on views/update-experimenter.jade:
```diff
@@ -0,0 +1,37 @@
+          .form-group
+            label.form-label.col-md-6(for='name') Name
+            .col-md-6
+              input#name.form-control(type='text', name='name', autofocus='')
```
The name and email text fields should probably be wider

jorgearanda commented:
It's getting close, Btara, thanks for your work!

btruhand commented:
I've done what you asked (or should have). Check out how it is now Jorge. I also separated the alert messages depending on the type of error now

jorgearanda commented:
Thanks, Btara!

jorgearanda commented:
BTW, Btara, please check my latest commit—moved some stuff around.
